Source/Dafny/Verifier/Translator.ExpressionWellformed.cs

```diff
@@ -0,0 +1,1295 @@
+        var correctConstructor = BplOr(e.LegalSourceConstructors.ConvertAll(
+          ctor => FunctionCall(e.tok, ctor.QueryField.FullSanitizedName, Bpl.Type.Bool, etran.TrExpr(e.Root))));
+        if (e.LegalSourceConstructors.Count == e.Type.AsDatatype.Ctors.Count) {
+          // Every constructor has this destructor; no need to check anything
+        } else {
+          builder.Add(Assert(GetToken(expr), correctConstructor,
+            new PODesc.ValidConstructorNames(DatatypeDestructor.PrintableCtorNameList(e.LegalSourceConstructors, "or"))));
+        }
+
+        CheckWellformedWithResult(e.ResolvedExpression, options, result, resultType, locals, builder, etran);
+        result = null;
+
+      } else if (expr is ConcreteSyntaxExpression) {
+        var e = (ConcreteSyntaxExpression)expr;
+        CheckWellformedWithResult(e.ResolvedExpression, options, result, resultType, locals, builder, etran);
+        result = null;
+
+      } else if (expr is BoogieFunctionCall) {
+        var e = (BoogieFunctionCall)expr;
+        foreach (var arg in e.Args) {
+          CheckWellformed(arg, options, locals, builder, etran);
+        }
+
+      } else {
+        Contract.Assert(false); throw new cce.UnreachableException();  // unexpected expression
+      }
+
+      if (result != null) {
+        Contract.Assert(resultType != null);
+        var bResult = etran.TrExpr(expr);
+        CheckSubrange(expr.tok, bResult, expr.Type, resultType, builder);
+        builder.Add(TrAssumeCmd(expr.tok, Bpl.Expr.Eq(result, bResult)));
+        builder.Add(TrAssumeCmd(expr.tok, CanCallAssumption(expr, etran)));
+        builder.Add(new CommentCmd("CheckWellformedWithResult: any expression"));
+        if (AlwaysUseHeap) {
+          builder.Add(TrAssumeCmd(expr.tok, MkIsAlloc(result, resultType, etran.HeapExpr)));
+        }
+        builder.Add(TrAssumeCmd(expr.tok, MkIs(result, resultType)));
+      }
+    }
+
+    private void CheckWellformedStmtExpr(StmtExpr stmtExpr, WFOptions options, Expr result, Type resultType, List<Variable> locals,
+      BoogieStmtListBuilder builder, ExpressionTranslator etran) {
+      // If we're inside an "old" expression, then "etran" will know how to translate
+      // expressions. However, here, we're also having to translate e.S, which is a
+      // Statement. Since statement translation (in particular, CallStmt's) works
+      // directly on the global variable $Heap, we change its value here.
+      // (Note, this would be a bad translation if the Statement could introduce
+      // control flow that would miss the "$Heap := tmpHeap;" assignment that
+      // resets the value of $Heap. Luckily, StmtExpr does not allow any such
+      // Statement.)
+
+      void addStatements() {
+        TrStmt(stmtExpr.S, builder, locals, etran);
+      }
+
+      void withFreshHeap(string heapVarSuffix, IToken token) {
```

Comment:
Maybe we could change this into:

```c#
BuildWithHeapAs(BoogieStmtListBuilder builder, 
  Bpl.Expr newHeap, 
  string heapStorageSuffix, 
  IToken token, 
  Action<BoogieStmtListBuilder> build) 
{
        var suffix = CurrentIdGenerator.FreshId(heapStorageSuffix);
        var currentHeapVar = new Bpl.LocalVariable(token, new Bpl.TypedIdent(token, "Heap$" + suffix, predef.HeapType));
        locals.Add(currentHeapVar);
        var tmpHeap = new Bpl.IdentifierExpr(token, currentHeapVar);
        var theHeap = new ExpressionTranslator(this, predef, token).HeapCastToIdentifierExpr;

        // currentHeap := $Heap;
        builder.Add(Bpl.Cmd.SimpleAssign(token, currentHeapVar, theHeap));
        // $Heap := newHeap;
        builder.Add(Bpl.Cmd.SimpleAssign(token, theHeap, newHeap));

        build(builder);

        // $Heap := currentHeap;
        builder.Add(Bpl.Cmd.SimpleAssign(token, theHeap, currentHeapVar));
}
```

And make that a top-level method.

Reply:
Good suggestions. Done.